fem/pfespace.cpp

```diff
@@ -3147,9 +3199,9 @@ static void SetSubVector(const int N,
                          const Array<int> &indices,
                          const Vector &in, Vector &out)
 {
-   auto y = out.Write();
-   const auto x = in.Read();
-   const auto I = indices.Read();
+   auto y = out.HostWrite();
```

Comment:
I realize this was the bug fix we were looking for for a long time, but it bothers me a little to do `SetSubVector` on the host in an object called `DeviceConformingProlongationOperator`. Shouldn't all the memory be on the device? Does this do a bunch of unnecessary copies?

Reply:
@barker29 Good question, and I think the answer is that the memory is already on host. `SetSubVector` is called only by `DeviceConformingProlongationOperator::BcastLocalCopy` and `DeviceConformingProlongationOperator::BcastEndCopy`, which are called only by `DeviceConformingProlongationOperator::Mult`, and when they are called, it is after some `HostRead` and `HostWrite` calls which are necessitated by `MPI_Isend` and `MPI_Irecv` calls. Thus the data is on host because of some `MPI` calls. An extra copy to host might hurt performance, but if the data is already on host, then setting a subvector on host seems fine to me.

Reply:
We should look at it as I remember that there can be direct `mpi` calls with data on the `device`.
I'm not sure that the `data` is always on the `host` there.

Reply:
It does not make sense to call `Host{Read,Write,ReadWrite}` and then use these host pointers inside `MFEM_FORALL` which expects device pointers. If this copy has to happen on host, then a normal `for` loop is what you need.

However, I do think we want to do these copies on device.

Reply:
One thing that is a bit tricky here is that only part of `y` is overwritten, so using `Write` may be incorrect if `y` is actually on host.

Reply:
@camierjs @v-dobrev Thanks for your comments, and I agree with them. First, I was not aware of `mpi_gpu_aware` (I guess @barker29 wrote this part?). This is interesting, and I did not know MPI calls could be made with data on device. It must be doing something more efficient than copying from device to host in this case. Is there some documentation on this? I only see a very concise declaration in `device.hpp`. For a few tests (unit tests and ex3p) with `pcuda` builds on lassen, I printed out `mpi_gpu_aware` and found it is always `false`. How can it be enabled? Assuming it can be `false` or `true`, I suppose the implementation of `SetSubVector` should depend on `mpi_gpu_aware` and `local` (if `local`, there is no MPI communication in any case). At least we can determine cases where data need not be on host.

Reply:
It can be enabled with `Device::SetGPUAwareMPI()`. The `MPI` implementation _should_ do something efficient, but we haven't seen yet one avoiding some internal copies (that was ~one year ago with `Laghos`) and could be revisited to see if the expected gain is reachable.

Reply:
@camierjs I tried adding `Device::SetGPUAwareMPI()` to ex3p, just before the declaration of `Device device`, but there is a segmentation fault, and TotalView shows the crash somewhere in the internals of `MPI_Isend`. Maybe I did something wrong (building configuration?).

Reply:
It has been a while since I tested this, and I cannot reproduce the bug that these `HostRead` calls fixed. I think the bug was only revealed in `MatrixFreeSolver` in branch `barker29/matrix-free-ams`, which seems to be working fine without `HostRead`. So I am reverting this change, and any issues can be fixed in a future PR for `barker29/matrix-free-ams`. It looks like there is a case that may need to be fixed, where `mpi_gpu_aware` and `local` are false in `DeviceConformingProlongationOperator::Mult` and the data is on host for the MPI communication. If so, the fix should not be in `SetSubVector` but in `DeviceConformingProlongationOperator::Mult`, with copying on host just in that case.